Add serial to shortcut list for policy

pearsonk opened the pull request:
Depends on https://github.com/OpenXT/vusb-daemon/pull/2

Will rebase when appropriate and dependency merged (assuming a merge commit will get shoved in the middle). Until then, can start some reviews at least.

Device serial numbers are otherwise only accessible via sysfs/udev property matches. With vendor/device IDs, having serial number as a shortcut makes it easier to lock particular devices (that supply a serial number) to individual VMs.

pearsonk committed:
Add serial to shortcut list for policy

Signed-off-by: Kevin Pearson <kevin.pearson.4@us.af.mil>

pearsonk commented:
Also updated device_lookup_by_attributes (which does not appear to ever be called) that had a parameter for serial and documentation for such, but was comparing to device shortname.

jean-edouard commented:
+1
Using the shortname as the "serial" is something the old deamon was doing that I picked up, but I agree it makes more sense to use the real serial, hoping most devices will have one.
The device lookup by attributes was used to implement the old always-assign configuration, in which the shortname was used as a serial.

jean-edouard commented:
#2 is merged. Github still seems happy about this pull request, but still shows the commit that's already in for some reason...
Maybe it's safer to rebase before merging?

pearsonk commented:
Yes, also I found a logic flaw in the always-assign that requires a quick 2-line fix. Currently the auto-assign does not apply serial to the newly-created rule (seems to start at ID=1000). The problem this causes is for multi-device systems such as having 2 of the same smart card readers, ends up with a precedent rule that assigns both to the first VM triggered. I will rebase this all but was waiting for the merge so I could do it all at once.

pearsonk commented on src/policy.c:
```diff
@@ -211,6 +213,8 @@ policy_set_sticky(int dev)
   new_rule->dev_vendorid = device->vendorid;
   new_rule->dev_deviceid = device->deviceid;
+  if (device->serial != NULL)
+    new_rule->dev_serial = device->serial;
```
Added check to policy_set_sticky to ensure that if present, a serial is set for newly-created rules for auto-passing a device to a specified VM.

pearsonk commented:
Rebased. The previous problem with the other commit being left was due to discrepancy created by introduction of a merge commit...

jean-edouard commented:
I built and did 3 tests:
- Added a deny rule for a specific serial, and made sure assignment failed. Success.
- Passed a device to a VM with "always assign" checked, then made sure the new rule contained the serial. Success.
- Plugged a random device that doesn't have a serial, and tried to assign it to a VM. Fail.

It looks like the condition you added in device_matches_rules doesn't work as expected. If the rule specifies a serial and the device doesn't have any, we want to return false, which doesn't seem to be the case.

pearsonk commented:
@jean-edouard Good catch, testing fix now

pearsonk commented on src/policy.c:
```diff
@@ -142,6 +142,16 @@ device_matches_rule(rule_t *rule, device_t *device)
+    /* Ensure the rule and device serials match */
+    else if (strcmp(rule->dev_serial,device->serial) != 0)
+      return false;
+  }
```
Modified this portion to include the check for when the rule states there should be a serial but the device does not have/present a serial

pearsonk commented:
Updated and commented the rebased change. Other logic was missing the (if rule.serial && !device.serial)

jean-edouard commented:
That seems to fix the issue. +1

pearsonk commented:
The removed line was a cut/paste casualty of me aligning the code order to be consistent with the rest of the code. Added the blank line back in and rebased.

And of course the rebase killed the comment, which GitHUb uses to make me seem less sane. For the purposes of history, @jean-edouard made a comment regarding a remove empty line that was later fixed.

jean-edouard commented:
@pearsonk thanks :)
Adding a commit would have kept the comment alive, but it's not ideal either...
I will merge this shortly.
